File: ui-testing/pages/inventory.py
from pages.base_page import BasePage
from selenium.webdriver.common.by import By


class Inventory(BasePage):
    SIDE = (By.ID, 'react-burger-menu-btn')
    ALL_ITEMS_LINK = (By.ID, 'inventory_sidebar_link')
    ABOUT_LINK = (By.ID, 'about_sidebar_link')
    LOGOUT_LINK = (By.ID, 'logout_sidebar_link')
    RESET_LINK = (By.ID, 'reset_sidebar_link')
    CLOSE_SIDEBAR = (By.ID, 'react-burger-cross-btn')
    SIDE_BAR = (By.CLASS_NAME, 'bm-menu-wrap')
    SHOP_CART = (By.CLASS_NAME, 'shopping_cart_link')
    CART_ITEM = (By.CLASS_NAME, 'shopping_cart_badge')

    ADD_ITEM_BACKPACK = (By.ID, 'add-to-cart-sauce-labs-backpack')
    ADD_ITEM_ONESIE = (By.ID, 'add-to-cart-sauce-labs-onesie')
    ADD_ITEM_RED_TSHIRT = (By.ID, 'add-to-cart-test.allthethings()-t-shirt-(red)')

    Inventory_List = (By.CLASS_NAME, 'inventory_list')
    ITEM_NAMES = (By.CLASS_NAME, 'inventory_item_name')
    SORT_SELECT = (By.CLASS_NAME, 'product_sort_container')
    SORT_ZA_OPTION = (By.CSS_SELECTOR, "option[value='za']")
    SORT_LOHI_OPTION = (By.CSS_SELECTOR, "option[value='lohi']")
    SORT_HILO_OPTION = (By.CSS_SELECTOR, "option[value='hilo']")
    ITEM_PRICE = (By.CLASS_NAME, 'inventory_item_price')

    def click_side(self, locator):
        element = self.driver.find_element(*locator)
        self.driver.execute_script("arguments[0].click();", element)

    def is_sidebar_hidden(self):
        status = self.driver.find_element(*self.SIDE_BAR).get_attribute('aria-hidden')
        print(status)

        return status == 'true'

    def click(self, locator):
        element = self.driver.find_element(*locator)
        self.driver.execute_script("arguments[0].click();", element)

    def get_all_items_name(self):
        """獲取所有商品名稱的文字清單"""
        # 使用 find_elements 抓取所有符合 ITEM_NAMES 的元素
        items = self.driver.find_elements(*self.ITEM_NAMES)
        # 加上 .strip() 確保去除換行或多餘空格，這對後續 A-Z 排序斷言非常重要
        name_list = [item.text.strip() for item in items]
        return name_list

    def click_sort_za(self):
        self.driver.find_element(*self.SORT_SELECT).click()
        self.driver.find_element(*self.SORT_ZA_OPTION).click()

    def click_sort_lohi(self):

        self.driver.find_element(*self.SORT_SELECT).click()
        self.driver.find_element(*self.SORT_LOHI_OPTION).click()

    def click_sort_hilo(self):

        self.driver.find_element(*self.SORT_SELECT).click()
        self.driver.find_element(*self.SORT_HILO_OPTION).click()

    def get_all_items_price(self):
        items = self.driver.find_elements(*self.ITEM_PRICE)
        price_list = [float(item.text.strip().replace('$', '')) for item in items]
        return price_list

File: playwright-testing/pages/inventory.ts
import { Page, Locator } from '@playwright/test';
import { BasePage } from './base.page';

export class InventoryPage extends BasePage {

  readonly sideMenuButton: Locator;
  readonly sidebarContainer: Locator;
  readonly allItemsLink: Locator;
  readonly aboutLink: Locator;
  readonly logoutLink: Locator;
  readonly resetLink: Locator;
  readonly closeSidebarButton: Locator;

  readonly shoppingCartLink: Locator;
  readonly shoppingCartBadge: Locator;

  readonly addBackpackButton: Locator;
  readonly addOnesieButton: Locator;
  readonly addRedTshirtButton: Locator;

  readonly inventoryList: Locator;
  readonly itemNames: Locator;
  readonly itemPrices: Locator;
  readonly sortSelect: Locator;

  constructor(page: Page) {
    super(page);


    this.sideMenuButton = page.locator('#react-burger-menu-btn');
    this.sidebarContainer = page.locator('.bm-menu-wrap');
    this.allItemsLink = page.locator('#inventory_sidebar_link');
    this.aboutLink = page.locator('#about_sidebar_link');
    this.logoutLink = page.locator('#logout_sidebar_link');
    this.resetLink = page.locator('#reset_sidebar_link');
    this.closeSidebarButton = page.locator('#react-burger-cross-btn');


    this.shoppingCartLink = page.locator('.shopping_cart_link');
    this.shoppingCartBadge = page.locator('.shopping_cart_badge');


    this.addBackpackButton = page.locator('#add-to-cart-sauce-labs-backpack');
    this.addOnesieButton = page.locator('#add-to-cart-sauce-labs-onesie');
    this.addRedTshirtButton = page.locator('[id="add-to-cart-test.allthethings()-t-shirt-(red)"]');


    this.inventoryList = page.locator('.inventory_list');
    this.itemNames = page.locator('.inventory_item_name');
    this.itemPrices = page.locator('.inventory_item_price');
    this.sortSelect = page.locator('.product_sort_container');
  }


  async openSideMenu(): Promise<void> {
    await this.sideMenuButton.click();
  }

  async closeSideMenu(): Promise<void> {
    await this.closeSidebarButton.click();
  }

  async clickLogout(): Promise<void> {
    await this.logoutLink.click();
  }

  async clickShoppingCart(): Promise<void> {
    await this.shoppingCartLink.click();
  }


  async isSidebarHidden(): Promise<boolean> {
    const status = await this.sidebarContainer.getAttribute('aria-hidden');
    return status === 'true';
  }


  async getAllItemNames(): Promise<string[]> {
    const items = await this.itemNames.allTextContents();
    return items.map(item => item.trim());
  }


  async getAllItemPrices(): Promise<number[]> {
    const items = await this.itemPrices.allTextContents();
    return items.map(item => parseFloat(item.replace('$', '').trim()));
  }


  async sortBy(optionValue: 'az' | 'za' | 'lohi' | 'hilo'): Promise<void> {
    await this.sortSelect.selectOption(optionValue);
  }
}